### ISSUE TITLE
No download link in customer account after purchases in 2.3.1

### ISSUE THREAD

**Issue by The-Krang** [magento/magento2#22390]
### Preconditions (*)

Hello again.  This is a continuance of the previous post that I put up, but it was closed.
Here is the link: https://github.com/magento/magento2/issues/22354

The problem still persists.

I guess I did not make the situation completely clear.
The $0.00 amounts in the tests were done with "Zero Subtotal Checkout", and the amounts **_above_** $0.00 were tested in Braintree and PayPal.

I am well aware that Paypal does not process amounts like $0.00. I have been using Magento 2 for years now.

Furthermore, the other night I downloaded version 2.3.1 from Bitnami, installed it, set it up like version 2.3.0, but there still are no download links, and the status still says “Pending”.

Version 2.3.0 works fine by providing download links, like the previous versions of Magento 2.x, but either something changed from version 2.3.0 to version 2.3.1, or I made a large mistake somewhere, or I have completely lost all comprehension.

Please take another look at what I presented.

Just to be even more clear, the **Live** purchases (**Production**, not sandbox) that I made through Braintree and Paypal using version 2.3.1 had the same outcome as the ones using "Zero Subtotal Checkout", and that is with the status showing "Pending" instead of "Available", and no download link.

The screen-shots I provided showed the "Zero Subtotal Checkout" tests that I did, but they also represent the **Live** purchases because the outcomes were the same.

Will anyone take a look at this?

Is there anymore information I can provide, like PHP, HTML, or JavaScript files?

What is a vanilla instance, and how do I get it so I can upload it or whatever?

**Comment by krishprakash** [comment]
Hi @The-Krang. Thank you for your report.
Please explain in proper issue format.

**Comment by The-Krang** [comment]
**With all due respect**, this process is exasperating and sloppy, and to be frank, silly.  I made a link https://github.com/magento/magento2/issues/22354 above for my post that was closed.  It appeared that the previous post was read without paying attention to the detail.  Now this one is closed too.  I don't have a lot of spare time to constantly **format** my posts just to have them closed over and over again.  Again, **all due respect**.  Best wishes to all.  Goodbye.

**Comment by eleventybillion** [comment]
> **With all due respect**, this process is exasperating and sloppy, and to be frank, silly. I made a link #22354 above for my post that was closed. It appeared that the previous post was read without paying attention to the detail. Now this one is closed too. I don't have a lot of spare time to constantly **format** my posts just to have them closed over and over again. Again, **all due respect**. Best wishes to all. Goodbye.

@The-Krang Did you manage to get your downloads working, I've just managed to sort today as I was experiencing the same issues since updating to 2.3.1.

**Comment by The-Krang** [comment]
@eleventybillion Unfortunately, no.  Thank you for responding to this issue.  Would you mind sharing with the rest of us how you managed to make it work for you?  Thank You.

**Comment by The-Krang** [comment]
Hello Mr. Thomas.
Thank you for contacting me on this issue.
Unfortunately, no. Would you mind letting me know how you managed to make it work for you? Thank You.

________________________________

**Comment by The-Krang** [comment]
To all who may also have this problem, I will post the fix I found, as soon as possible.

**Comment by The-Krang** [comment]
I knew that I had not lost my comprehension.  I was correct, there was a change somewhere from Magento 2.3.0 to 2.3.1 that affected download links.


This is what I did to fix my problem.

I did some more searching, and what I found was that on my system, the file with the problem was this one: 

/public_html/vendor/magento/module-downloadable/Observer/SaveDownloadableOrderItemObserver.php

I renamed this problem file to keep it as a backup just in case I needed to go back to it.

I made a replacement SaveDownloadableOrderItemObserver.php file from the code from the link **BELOW**.


Credit goes to James/crankycyclops, for his bug report and the fix for the bug.

The code that fixed my issure came from this page:

https://github.com/magento/magento2/blob/5410e077c534127f3492d53f8ab3981691bed2e3/app/code/Magento/Downloadable/Observer/SaveDownloadableOrderItemObserver.php


The previous orders **BEFORE** the fix stayed the same, and the newest orders work like they should and did in all versions prior to Magento 2.3.1.


I made sure that these settings were in place in my Admin:

Under stores> configuration> catalog> catalog> Downloadable Product Options> I have "Order Item Status to Enable Downloads" set to "Pending".

Under stores> configuration> catalog> Inventory> for both "Stock Options" and "Product Stock Options", I have "Decrease Stock When Order is Placed" and "Manage Stock" respectively set to "No".

Under stores> configuration> sales> payment methods> Zero Subtotal Checkout> I have "New Order Status" set to "Pending".

Under stores> configuration> sales> payment methods> PayPal Express Checkout> Basic Settings - PayPal Express Checkout> I have "Payment Action" set to "Sale".  This was a Live test, that is, money was transfered.


That is all.
